src/classes/ADDR_CannotDelete_TDTM.cls

```diff
@@ -0,0 +1,90 @@
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2020
+* @group Addresses
+* @group-content ../../ApexDocContent/Addresses.htm
+* @description Stops an Address from being deleted if it has any Account or Contact child records. 
+*/
+public with sharing class ADDR_CannotDelete_TDTM extends TDTM_Runnable {
+    /*******************************************************************************************************
+    * @description Get the setting of preventing Address deletion
+    */
+    private static Boolean enabledPreventAddressDeletion = UTIL_CustomSettingsFacade.getSettings().Prevent_Address_Deletion__c;
+    
+    /*******************************************************************************************************
+    * @description Stops an Address from being deleted if it has any Account or Contact child records. 
+    * @param newlist the list of Addresses from trigger new. 
+    * @param oldlist the list of Addresses from trigger old. 
+    * @param triggerAction which trigger event (BeforeInsert, AfterInsert, etc.). 
+    * @param objResult the describe for Addresses 
+    * @return DmlWrapper.  
+    ********************************************************************************************************/
+    public override DmlWrapper run(List<SObject> newlist, List<SObject> oldlist, 
+    TDTM_Runnable.Action triggerAction, Schema.DescribeSObjectResult objResult) {
+        
+        if (!enabledPreventAddressDeletion) {
+            return new DmlWrapper(); 
+        }
+        
+        Map<ID, Address__c> oldmap = new Map<ID, Address__c>((List<Address__c>)oldList);
+        
+        if (triggerAction == TDTM_Runnable.Action.BeforeDelete) {
+            for (Address__c addr : [
+                                    SELECT Id, 
+                                        (SELECT Id FROM Address__c.Accounts__r LIMIT 1),
+                                        (SELECT Id FROM Address__c.Contacts1__r LIMIT 1)
+                                    FROM Address__c WHERE Id IN :oldlist
+                                    
+                                ]) {
+                
+                if (this.hasChildRecords(addr)) {
+                    Address__c addressInContext = oldMap.get(addr.ID);
+                    addressInContext.addError(Label.CannotDelete);
+                }
+            }     
+        }
+
+        return new DmlWrapper();
+    }
+    
+    /*******************************************************************************************************
+     * @description Evaluates whether the Address has any child (Account or Contact) related records.
+     * @param addr is the current Address record.
+     * @return Boolean.
+     ********************************************************************************************************/
+    @testVisible
+    private Boolean hasChildRecords(Address__c addr) {
```

Comment:
@MatthewBlanski something out of the scope of this PR, since we are/will be using hasChildRecords method in prevent delete PRs, should we consider moving this to its own class with switch or is it just an over kill?

Reply:
@skosgi1 For now, let's keep this inside this class. This would be a good candidate to move to an address service later downstream.